modules/images/webp-uploads/load.php

```diff
@@ -683,3 +683,190 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
+ *
+ * @since n.e.x.t
+ *
+ * @param int   $attachment_id The ID representing the attachment.
+ * @param array $data The current metadata of the attachment.
+ * @return array The updated metadata for the attachment.
+ */
+function webp_uploads_backup_sources( $attachment_id, $data ) {
+	if ( ! isset( $data['sources'] ) ) {
+		return $data;
+	}
+
+	$data['_sources'] = $data['sources'];
+	// Remove the current sources as at this point the current values are no longer accurate.
+	unset( $data['sources'] );
+
+	return $data;
+}
+
+/**
+ * Restore an image from the backup sizes, the current hook moves the `sources` from the `full-orig` key into
+ * the top level `sources` into the metadata, in order to ensure the restore process has a reference to the right
+ * images. When `IMAGE_EDIT_OVERWRITE` is defined and is truthy the function would loop into the sources array
+ * and remove any sources that was created as an edited version, the filenames for those cases are constructed out
+ * of `e-{digits}` where `{digits}` is a number of length 13.
+ *
+ * @param int   $attachment_id The ID of the attachment.
+ * @param array $data The current metadata to be stored in the attachment.
+ * @return array The updated metadata of the attachment.
+ */
+function webp_uploads_restore_image( $attachment_id, $data ) {
+	$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+
+	if ( ! is_array( $backup_sizes ) ) {
+		return $data;
+	}
+
+	// If `IMAGE_EDIT_OVERWRITE` is defined and is truthy remove any edited images if present before replacing the metadata.
+	if ( defined( 'IMAGE_EDIT_OVERWRITE' ) && IMAGE_EDIT_OVERWRITE ) {
+		$file     = get_attached_file( $attachment_id );
+		$dirname  = pathinfo( $file, PATHINFO_DIRNAME );
+		$metadata = wp_get_attachment_metadata( $attachment_id );
+
+		$sources = isset( $metadata['sources'] ) && is_array( $metadata['sources'] ) ? $metadata['sources'] : array();
+
+		foreach ( $sources as $mime => $properties ) {
+			if ( empty( $properties['file'] ) ) {
+				continue;
+			}
+
+			// Delete only if it's an edited image.
+			if ( preg_match( '/-e\d{13}/', $properties['file'] ) ) {
+				$delete_file = path_join( $dirname, $properties['file'] );
+				wp_delete_file( $delete_file );
+			}
+		}
+
+		foreach ( $metadata['sizes'] as $size_name => $properties ) {
+			if ( ! isset( $properties['sources'] ) || ! is_array( $properties['sources'] ) ) {
+				continue;
+			}
+
+			foreach ( $properties['sources'] as $mime => $source_properties ) {
+				if ( empty( $source_properties['file'] ) ) {
+					continue;
+				}
+
+				// Delete only if it's an edited image.
+				if ( preg_match( '/-e\d{13}/', $source_properties['file'] ) ) {
+					$delete_file = path_join( $dirname, $source_properties['file'] );
+					wp_delete_file( $delete_file );
+				}
+			}
+		}
+	}
+
+	if ( isset( $backup_sizes['full-orig']['sources'] ) ) {
+		$data['sources'] = $backup_sizes['full-orig']['sources'];
+	}
```

Comment:
This is the only part of this function which we actually need. In `webp_uploads_backup_sources()` we (rightfully) only back up the main `sources`, so here we have to restore them. The `sources` for every image size should work out of the box with WP core, since it just copies/moves over the complete size data, which would include `sources`. Again, we don't generate additional sources for image edits, so no need to delete any files here.

On another note though, we can't rely on `$backup_sizes['full-orig']['sources']` as that won't be set. When you review the code from WP core's `wp_save_image()` and `wp_restore_image()` closely, you will see that they manually assemble the `full-orig` array, which means it'll never include the `sources`. Only for the sub-sizes it copies/moves over the original data, so the `sources` there are maintained.

In other words, what we need to do here is rely on our custom storage (e.g. the `_wp_attachment_backup_sources` meta key I proposed above, or your current implementation `_sources`).

Reply:
> On another note though, we can't rely on $backup_sizes['full-orig']['sources'] as that won't be set. When you review the code from WP core's wp_save_image() and wp_restore_image() closely, you will see that they manually assemble the full-orig array, which means it'll never include the sources. Only for the sub-sizes it copies/moves over the original data, so the sources there are maintained.

Right this logic is based on the fact that we ensure there's always a `sources` attributes for `full-orig` image. 

> In other words, what we need to do here is rely on our custom storage (e.g. the _wp_attachment_backup_sources meta key I proposed above, or your current implementation _sources).

Great let me take a look at this approach instead which would mean this section of code need to be updated.